### DISCUSSION TITLE
How to properly use XLIFF with DjangoCMS

### DISCUSSION THREAD

**Discussion by vasekch** [energie360/djangocms-xliff#5] [General]
Hi, I'm trying to get my head around how to use this.

I'm familiar with i18n in Django and I can use gettext `.po` files, compile it and reload server to get my updated translations.

But from the `README.md`, I'm a bit confused.

> With djangocms-xliff it is possible to export all text objects from a page into an XLIFF-compatible file and re-import the file at the end of the translation process.

What are the text objects? Is this from `Plugins`, `PageExtensions` or `TitleExtensions`? Are gettext messages included?

> Import the XLIFF to the same page in the same language you exported from with Languages > Import from XLIFF

It seems a bit unhandy to export each page separately? Can I have complete website exported for translation?

> You can only import the file to the same page and language that you exported from before.
> It is not possible to export a file for one language and import it to another language.

How can I use this to support translating my site in a completely new language?

Maybe I'm missing something similar to base language concept of the gettext tooling.

Could you please elaborate a bit? Thank you.

**Comment by andre-dasilva** [comment]
Hi @vasekch 

Thank you for your question. Glad you have some interest in the project 😉 

Here the answers:

> What are the text objects? Is this from Plugins, PageExtensions or TitleExtensions? 

Currently the following gets exported / imported:
* Page (more like Title) metadata fields:
  * title, slug, menu_title, page_title, meta_description 
* All Plugin (model) fields on a Page, that are of type CharField, SlugField, TextField or URLField. 
 
Extensions are not supported, but could be in the future.

> Are gettext messages included?

No, for now its just CMS Page / Title Metadata and Plugins (and in the future custom Models)

> It seems a bit unhandy to export each page separately? Can I have complete website exported for translation?

Currently only single page import / exports are possible. The feature to export all pages at once, could be written with the existing functions (if someone is interested, take a look at: https://github.com/energie360/djangocms-xliff/blob/main/djangocms_xliff/exports.py#L23. You could loop through all CMS pages and use them with this function)

> How can I use this to support translating my site in a completely new language?

Example you want to translate German to French.

Then the flow in the CMS should be like this:
1. Create your page in german with all your plugins
2. Create the french page 
3. Copy all plugins from the german page to the french page
4. On the french page: "Export as XLIFF"
5. Send the XLIFF file to your agency or translate them yourself
6. On the french page "Import as XLIFF"

The flow is needed because the plugins on a Page in the django-cms are completly seperate from each other. 
So a plugin on the german site does not know anything about the same plugin on the french page. 
Maybe this will be supported in the future by the django-cms. But for now this flow is needed because otherwise you loose the plugin id and don't have save way to export / import the plugins.   

> Maybe I'm missing something similar to base language concept of the gettext tooling.

No you are not missing something 😃 . Gettext is more of a developer tool to translate stuff. You run your command and send .po the file to your agency get the file back and you're done. This package was intended for non developers, so that they can import / export pages themselves. Because of the limitation on the cms, the "complicated" flow exists.

Feel free to ask more questions, or look at the source code 😃
